lxc/storage.go

```diff
@@ -0,0 +1,861 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/olekukonko/tablewriter"
+	"gopkg.in/yaml.v2"
+
+	"github.com/lxc/lxd"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/i18n"
+	"github.com/lxc/lxd/shared/termios"
+)
+
+// This command is only allowed to create custom storage volumes.
+const defaultStoragePoolVolumeType = "custom"
+
+type storageCmd struct {
+}
+
+func (c *storageCmd) showByDefault() bool {
+	return true
+}
+
+func (c *storageCmd) storagePoolEditHelp() string {
+	return i18n.G(
+		`### This is a yaml representation of a storage pool.
+### Any line starting with a '# will be ignored.
+###
+### A storage pool consists of a set of configuration items.
+###
+### An example would look like:
+### name: default
+### driver: zfs
+### used_by: []
+### config:
+###   size: "61203283968"
+###   source: /home/chb/mnt/lxd_test/default.img
+###   zfs.pool_name: default`)
+}
+
+func (c *storageCmd) storagePoolVolumeEditHelp() string {
+	return i18n.G(
+		`### This is a yaml representation of a storage volume.
+### Any line starting with a '# will be ignored.
+###
+### A storage volume consists of a set of configuration items.
+###
+### name: vol1
+### type: custom
+### used_by: []
+### config:
+###   size: "61203283968"`)
+}
+
+func (c *storageCmd) usage() string {
+	return i18n.G(
+		`Manage storage.
+
+lxc storage list [<remote>:]                           List available storage pools.
+lxc storage show [<remote>:]<pool>                     Show details of a storage pool.
+lxc storage create [<remote>:]<pool> [key=value]...    Create a storage pool.
+lxc storage get [<remote>:]<pool> <key>                Get storage pool configuration.
+lxc storage set [<remote>:]<pool> <key> <value>        Set storage pool configuration.
+lxc storage unset [<remote>:]<pool> <key>              Unset storage pool configuration.
+lxc storage delete [<remote>:]<pool>                   Delete a storage pool.
+lxc storage edit [<remote>:]<pool>
+    Edit storage pool, either by launching external editor or reading STDIN.
+    Example: lxc storage edit [<remote>:]<pool> # launch editor
+             cat pool.yaml | lxc storage edit [<remote>:]<pool> # read from pool.yaml
+
+lxc storage attach [<remote>:]<pool> <container> [device name]
+lxc storage attach-profile [<remote:>]<pool> <profile> [device name]
+
+lxc storage detach <pool> <container> [device name]
+lxc storage detach-profile <pool> <container> [device name]
```

Comment:
For now I think we should drop "attach", "attach-profile", "detach" and "detach-profile".

"attach" and "detach" don't really make sense at the pool level since you can't change the backing pool of a container. The profile ones may work but will similarly fail if you have any container using the profile since again, you can't unset the rootfs.